modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java

```diff
@@ -29,25 +30,54 @@ public class MongoDBContainer extends GenericContainer<MongoDBContainer> {
 
     private static final String MONGODB_DATABASE_NAME_DEFAULT = "test";
 
+    private static final String STARTER_SCRIPT = "/testcontainers_start.sh";
+
+    private final boolean shardingEnabled;
+
     /**
      * @deprecated use {@link MongoDBContainer(DockerImageName)} instead
      */
     @Deprecated
     public MongoDBContainer() {
-        this(DEFAULT_IMAGE_NAME.withTag(DEFAULT_TAG));
+        this(DEFAULT_IMAGE_NAME.withTag(DEFAULT_TAG), false);
     }
 
     public MongoDBContainer(@NonNull final String dockerImageName) {
-        this(DockerImageName.parse(dockerImageName));
+        this(DockerImageName.parse(dockerImageName), false);
     }
 
     public MongoDBContainer(final DockerImageName dockerImageName) {
+        this(dockerImageName, false);
+    }
+
+    public MongoDBContainer(final DockerImageName dockerImageName, boolean shardingEnabled) {
         super(dockerImageName);
+        this.shardingEnabled = shardingEnabled;
         dockerImageName.assertCompatibleWith(DEFAULT_IMAGE_NAME);
 
         withExposedPorts(MONGODB_INTERNAL_PORT);
-        withCommand("--replSet", "docker-rs");
-        waitingFor(Wait.forLogMessage("(?i).*waiting for connections.*", 1));
+        if (shardingEnabled) {
+            withCreateContainerCmdModifier(cmd -> {
+                cmd.withEntrypoint("sh");
+            });
+            withCommand("-c", "while [ ! -f " + STARTER_SCRIPT + " ]; do sleep 0.1; done; " + STARTER_SCRIPT);
+            waitingFor(Wait.forLogMessage("(?i).*mongos ready.*", 1));
+        } else {
+            withCommand("--replSet", "docker-rs");
+            waitingFor(Wait.forLogMessage("(?i).*waiting for connections.*", 1));
+        }
+    }
+
+    @Override
+    protected void containerIsStarting(InspectContainerResponse containerInfo) {
+        copyFileToContainer(MountableFile.forClasspathResource("/sharding.sh", 0777), STARTER_SCRIPT);
```

Comment:
Let's copy `sharding.sh` only when `shardingEnabled` is `true`

```suggestion
        if (this.shardingEnabled) {
            copyFileToContainer(MountableFile.forClasspathResource("/sharding.sh", 0777), STARTER_SCRIPT);
        }
```